modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -2267,35 +2284,242 @@ public int compare(CodegenProperty one, CodegenProperty another) {
+        Collections.sort(modelNames);
+        return modelNames;
+    }
+
     protected CodegenDiscriminator createDiscriminator(String schemaName, Schema schema) {
-        if (schema.getDiscriminator() == null) {
+        Discriminator sourceDiscriminator = recursiveGetDiscriminator(schema);
+        if (sourceDiscriminator == null) {
             return null;
         }
         CodegenDiscriminator discriminator = new CodegenDiscriminator();
-        discriminator.setPropertyName(toVarName(schema.getDiscriminator().getPropertyName()));
-        discriminator.setPropertyBaseName(schema.getDiscriminator().getPropertyName());
+        String discPropName = sourceDiscriminator.getPropertyName();
+        discriminator.setPropertyName(toVarName(discPropName));
+        discriminator.setPropertyBaseName(sourceDiscriminator.getPropertyName());
         discriminator.setPropertyGetter(toGetter(discriminator.getPropertyName()));
         // FIXME: for now, we assume that the discriminator property is String
         discriminator.setPropertyType(typeMapping.get("string"));
-        discriminator.setMapping(schema.getDiscriminator().getMapping());
-        if (schema.getDiscriminator().getMapping() != null && !schema.getDiscriminator().getMapping().isEmpty()) {
-            for (Entry<String, String> e : schema.getDiscriminator().getMapping().entrySet()) {
+        discriminator.setMapping(sourceDiscriminator.getMapping());
+        if (sourceDiscriminator.getMapping() != null && !sourceDiscriminator.getMapping().isEmpty()) {
+            for (Entry<String, String> e : sourceDiscriminator.getMapping().entrySet()) {
                 String nameOrRef = e.getValue();
                 String name = nameOrRef.indexOf('/') >= 0 ? ModelUtils.getSimpleRef(nameOrRef) : nameOrRef;
                 String modelName = toModelName(name);
                 discriminator.getMappedModels().add(new MappedModel(e.getKey(), modelName));
             }
         } else {
-            Map<String, Schema> allDefinitions = ModelUtils.getSchemas(this.openAPI);
-            allDefinitions.forEach((childName, child) -> {
-                if (child instanceof ComposedSchema && ((ComposedSchema) child).getAllOf() != null) {
-
-                    final List<String> parentSchemas = ModelUtils.getAllParentsName((ComposedSchema) child, allDefinitions, true);
-                    if (parentSchemas.contains(schemaName)) {
-                        discriminator.getMappedModels().add(new MappedModel(childName, toModelName(childName)));
-                    }
+            // for schemas that allOf inherit from this schema, add those descendents to this discriminator map
+            List<String> allOfDescendents = getDescendentSchemaNames(schemaName);
+            for (String allOfDescendent: allOfDescendents) {
+                discriminator.getMappedModels().add(new MappedModel(allOfDescendent, toModelName(allOfDescendent)));
```

Comment:
We are adding more models to our discriminator map here than we have in the past.
Should we add those models into the models import where this is called in fromModel?